notifier/notifier.go

```diff
@@ -618,10 +654,27 @@ func (n *Manager) sendOne(ctx context.Context, c *http.Client, url string, b []b
 	return nil
 }
 
-// Stop shuts down the notification handler.
+// Stop signals the notification manager to shut down.
+//
+// Stop will optionally wait up to the configured drain timeout to drain any queued notifications
+// before shutting down. Any queued notifications remaining at the end of the timeout will be dropped.
+//
+// A drain timeout of 0 means that no queued notifications will be drained and they will instead be dropped.
 func (n *Manager) Stop() {
+	defer n.cancel()
+
 	level.Info(n.logger).Log("msg", "Stopping notification manager...")
-	n.cancel()
+	n.stopRequested <- struct{}{}
+
+	if n.opts.DrainTimeout == 0 {
+		// Draining disabled, we are done.
+		level.Info(n.logger).Log("msg", "Notification manager stopped")
+		return
+	}
+
+	<-n.stopped
+	n.drainQueue()
```

Comment:
I'm not convinced it's safe to drain the queue here. As you stated in another comment, the `Run()` contract is that it blocks until the Manager has stopped executing. Moving the draining here is breaking such contract.